pulpcore/app/models/publication.py

```
@@ -553,3 +556,42 @@ def invalidate_cache(self):
         if settings.CACHE_ENABLED:
             Cache().delete(base_key=self.base_path)
             # Can also preload cache here possibly
+
+
+class ArtifactDistribution(Distribution):
+    """Serve artifacts by their sha256.
+    This is in tech-preview.
+    """
+
+    TYPE = "artifact"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        from pulpcore.app.serializers import ContentRedirectContentGuardSerializer
+
+        self.content_guard = ContentRedirectContentGuardSerializer.get_or_create(
+            {"name": "artifact redirect"}
+        )
+
+    def artifact_url(self, artifact):
+        origin = settings.CONTENT_ORIGIN.strip("/")
+        prefix = settings.CONTENT_PATH_PREFIX.strip("/")
+        base_path = self.base_path.strip("/")
+
+        plain_url = urljoin(urljoin(urljoin(origin, prefix + "/"), base_path + "/"), str(artifact.pk))
+        return self.content_guard.cast().preauthenticate_url(plain_url)
+
+    def content_handler(self, path):
+        """Serve an artifact or return 404."""
+        uuid = path.rstrip("/")
+        if re.match(r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}", uuid):
+            artifact = Artifact.objects.get(pk=uuid)
+            if artifact:
+                return ArtifactResponse(artifact)
+        return HTTPNotFound()
```

Please raise this exception.